src/node/ledger_secrets.h

```diff
@@ -3,118 +3,283 @@
-#include <nlohmann/json.hpp>
+#include <map>
+#include <optional>
 
 namespace ccf
 {
   struct LedgerSecret
   {
-    static constexpr auto MASTER_KEY_SIZE = crypto::GCM_SIZE_KEY;
-
-    std::vector<uint8_t> master; // Referred to as "sd" in TR
+    std::vector<uint8_t> raw_key;
+    std::shared_ptr<crypto::KeyAesGcm> key;
 
     bool operator==(const LedgerSecret& other) const
     {
-      return master == other.master;
+      return raw_key == other.raw_key;
     }
 
-    LedgerSecret()
+    LedgerSecret() = default;
+
+    LedgerSecret(const LedgerSecret& other) :
+      raw_key(other.raw_key),
+      key(std::make_shared<crypto::KeyAesGcm>(other.raw_key))
+    {}
+
+    LedgerSecret(std::vector<uint8_t>&& raw_key_) :
+      raw_key(raw_key_),
+      key(std::make_shared<crypto::KeyAesGcm>(std::move(raw_key_)))
+    {}
+  };
+
+  inline LedgerSecret make_ledger_secret()
+  {
+    return LedgerSecret(tls::create_entropy()->random(crypto::GCM_SIZE_KEY));
+  }
+
+  using VersionedLedgerSecret = std::pair<kv::Version, LedgerSecret>;
+  using LedgerSecretsMap = std::map<kv::Version, LedgerSecret>;
+
+  struct VersionedLedgerSecrets
+  {
+    LedgerSecretsMap secrets;
+
+    VersionedLedgerSecrets() = default;
+
+    VersionedLedgerSecrets(LedgerSecretsMap&& secrets_) :
+      secrets(std::move(secrets_))
+    {}
+
+    bool operator==(const VersionedLedgerSecrets& other) const
     {
-      master = tls::create_entropy()->random(MASTER_KEY_SIZE);
+      return secrets == other.secrets;
     }
-
-    LedgerSecret(const std::vector<uint8_t>& master_) : master(master_) {}
   };
 
   class LedgerSecrets
   {
-  public:
-    struct VersionedLedgerSecret
+  private:
+    Secrets& secrets_table;
+
+    std::optional<NodeId> self = std::nullopt;
+
+    SpinLock lock;
+    VersionedLedgerSecrets ledger_secrets;
+    std::optional<LedgerSecretsMap::iterator> commit_key_it = std::nullopt;
+
+    LedgerSecretsMap::iterator get_encryption_key_it(kv::Version version)
     {
-      kv::Version version;
-      LedgerSecret secret;
+      // Encryption key for a given version is the one with the highest version
+      // that is lower than the given version (e.g. if encryption_keys contains
+      // two keys for version 0 and 10 then the key associated with version 0
+      // is used for version [0..9] and version 10 for versions 10+)
 
-      bool operator==(const VersionedLedgerSecret& other) const
+      auto search_begin =
+        commit_key_it.value_or(ledger_secrets.secrets.begin());
+
+      auto search = std::upper_bound(
+        search_begin,
+        ledger_secrets.secrets.end(),
+        version,
+        [](auto a, const auto& b) { return b.first > a; });
+      if (search == search_begin)
+      {
+        throw std::logic_error(fmt::format(
+          "kv::TxEncryptor: could not find ledger encryption key for seqno {}",
+          version));
+      }
+      return --search;
+    }
+
+    void take_dependency_on_secrets(kv::Tx& tx)
```

Comment:
I think this should be `ReadOnlyTx` and `get_read_only_view`. We don't use this in many places but if we did, we could get `const`-like type-safety by using `ReadOnlyTx` throughout some execution stacks.

Reply:
Good point. Changed now.